### ISSUE TITLE
Add on_close callback for GraphQLWebSocket

### ISSUE THREAD

**Issue by meh** [async-graphql/async-graphql#1022]
## Description of the feature

I'd like to be able to do something when the websocket connection for subscription closes, I can't find a way to do that through the axum websocket (I'd have to match on messages received, but I can't because the whole socket becomes owned by async-graphql), and can't find anything on the GraphQLWebSocket itself.

## Code example (if possible)

```rust
GraphQLWebSocket::new(socket, schema, protocol)
  .on_close(|ctx| async move { ... })
```

**Comment by sunli829** [comment]
Do it like below:

```rust
async fn graphql_subscription(
    ws: WebSocketUpgrade,
    schema: Extension<BooksSchema>,
    protocol: GraphQLProtocol,
) -> impl IntoResponse {
    ws.protocols(ALL_WEBSOCKET_PROTOCOLS)
        .on_upgrade(move |socket| {
            println!("connection open");
            let resp = GraphQLWebSocket::new(socket, schema.0, protocol).serve();
            println!("connection close");
            resp
        })
}
```

**Comment by meh** [comment]
I guess I shuold `await` on the `serve()` call?

**Comment by sunli829** [comment]
My fault, this is the correct example. 😂

```rust
async fn graphql_subscription(
    ws: WebSocketUpgrade,
    schema: Extension<BooksSchema>,
    protocol: GraphQLProtocol,
) -> impl IntoResponse {
    ws.protocols(ALL_WEBSOCKET_PROTOCOLS)
        .on_upgrade(move |socket| async move {
            println!("connection open");
            GraphQLWebSocket::new(socket, schema.0, protocol)
                .serve()
                .await;
            println!("connection close");
        })
}
```

**Comment by meh** [comment]
Also is there a reason `on_connection_init` takes an `Fn` and not an `FnOnce`?

**Comment by sunli829** [comment]
It's a bug. 😂

**Comment by sunli829** [comment]
Fixed in `v4.0.9`

**Comment by meh** [comment]
Thanks! Everything works now :DDD
